### PR TITLE
Place dependencies into separate file

### PR DESCRIPTION
**Pull request by K20shores**
This separate outs the dependencies and partially addresses #2713

- All `find_package` calls are now in the `dependencies.cmake` file, making it easyish to see in one place what is needed
- Added a file to start to collect the macros and functions as recommended by the linked issue

### CODE REVIEW

**Comment by ZedThree** [comment]
Probably quite a lot of this could also be moved into separate `Find*.cmake` files to tidy up even further

File: CMakeLists.txt [hunk 1]
```diff
@@ -1538,8 +927,6 @@ IF(ENABLE_PNETCDF)
   # Check for ncmpi_create in libpnetcdf, define USE_PNETCDF
   # Does the user want to turn on PnetCDF read ability?
   SET(USE_PNETCDF ON CACHE BOOL "")
-  FIND_LIBRARY(PNETCDF NAMES pnetcdf)
-  FIND_PATH(PNETCDF_INCLUDE_DIR pnetcdf.h)
   IF(NOT PNETCDF)
```
Comment (ZedThree) [inline CMakeLists.txt:930]: This now isn't set until the `include(cmake/dependencies.cmake)` line later, so pnetcdf is reported as always disabled (probably requesting a feature and not finding it should be a hard error, which would catch this). I'm not sure if it's as simple as moving that earlier, given there's potentially some interaction between options (another good reason to try and disentangle them all!).

I've whipped up a very simple `FindPNETCDF` module that might be able to sort some of this out. I can make a PR into this branch if you like

Comment (K20shores) [reply CMakeLists.txt:940]: @ZedThree I've just moved it. I wasn't able to get the tests to pass for the parallel version on my machine but I was getting a different error than the CI anyways. Feel free to make that PR! My goal was to turn as many of the dependencies as possible into `Find*` packages.

**Comment by K20shores** [comment]
@WardF after @ZedThree's suggestion, the tests pass and it appears this PR is ready to go.

**Comment by WardF** [comment]
Thanks @K20shores and @ZedThree.  Hope everybody stayed/is staying warm.  Now that things are thawing out in rural Oregon, I should be able to resume review of the PR backlog.

**Comment by WardF** [comment]
One last change occurs to me; the new file needs to be added to the`EXTRA_DIST` definition in the `makefile.am` file in the corresponding directory. This way, when we create tarballs with `make dist` (using an autotools-based build), the tarball will have all the files needed for cmake-based builds.

**Comment by WardF** [comment]
I can make that change; other than that, this looks great, thanks!

**Comment by K20shores** [comment]
@WardF `Makefile.am` seems to include the `cmake` directory; that's where the `dependencies.cmake` and `netcdf_functions_macros.cmake` file are. Does it still need to be added to `Makefile.am`?

**Comment by WardF** [comment]
If it includes the directory we should be good. I'll double check to be sure, but that's probably got it. Thanks!

**Pull request merged**